<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Lecture 8: Simple model of evolution and introduction to sequence alignment &#8212; bioinformatics@nau</title>
    
    <link rel="stylesheet" href="../../../static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="../../../static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../static/jquery.js"></script>
    <script type="text/javascript" src="../../../static/underscore.js"></script>
    <script type="text/javascript" src="../../../static/doctools.js"></script>
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Lecture 9: Sequence alignment algorithms" href="lecture9.html" />
    <link rel="prev" title="Lecture 7: Molecular evolution and introduction to sequence searching and phylogenetic inference" href="lecture7.html" /> 
  </head>
  <body role="document">
      <div class="header" role="banner"><h1 class="heading"><a href="../../../index.html">
          <span>bioinformatics@nau</span></a></h1>
        <h2 class="heading"><span>Lecture 8: Simple model of evolution and introduction to sequence alignment</span></h2>
      </div>
      <div class="topnav" role="navigation" aria-label="top navigation">
      
        <p>
        «&#160;&#160;<a href="lecture7.html">Lecture 7: Molecular evolution and introduction to sequence searching and phylogenetic inference</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="lecture9.html">Lecture 9: Sequence alignment algorithms</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="lecture-8-simple-model-of-evolution-and-introduction-to-sequence-alignment">
<h1>Lecture 8: Simple model of evolution and introduction to sequence alignment<a class="headerlink" href="#lecture-8-simple-model-of-evolution-and-introduction-to-sequence-alignment" title="Permalink to this headline">¶</a></h1>
<dl class="docutils">
<dt>Sequence alignment: each position in a sequence is treated as an independent character state. Alignments represent a hypothesis about homology of the sequences themselves, and positional homology across the sequences. Alignments are used in generating input for phylogenetic tree construction and database searching.</dt>
<dd><ul class="first last simple">
<li>Local versus global alignments</li>
<li>Primary structure alignments: De novo alignment, Reference-based alignment</li>
<li>Secondary/tertiary structure alignments: Structure-based alignment</li>
</ul>
</dd>
<dt>Challenges in Phylogenetic Inference:</dt>
<dd><ul class="first last simple">
<li>homoplasy: occurrence of similarities not due to common ancestry (convergent and parallel evolution)</li>
<li>unequal rates of evolution: mutations may occur more in certain branches of &#8220;the tree&#8221;</li>
<li>nonvertical evolution: a branching tree is often an oversimplification of actual evolutionary processes (e.g., due to horizontal gene transfer)</li>
</ul>
</dd>
<dt>Why molecular sequence data rather than morphological data?</dt>
<dd><ul class="first last simple">
<li>cost of sequencing has dropped: now the cheapest data to collect for inferring phylogeny</li>
<li>discrete character states (4 nucleotides, 20 amino acids) makes molecular sequence data more straight-forward to work with than morphological data</li>
</ul>
</dd>
</dl>
<div class="section" id="script-to-model-sequence-evolution">
<span id="sequence-evolution-script"></span><h2>Script to model sequence evolution<a class="headerlink" href="#script-to-model-sequence-evolution" title="Permalink to this headline">¶</a></h2>
<p>Here we&#8217;ll apply some of the above techniques to create a very basic model sequence evolution. You can copy and paste this code to run it on your own:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># import some functions from python&#39;s random module - these will</span>
<span class="c1"># be used in the modeling process</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="k">import</span> <span class="n">choice</span><span class="p">,</span> <span class="n">randint</span>
<span class="c1"># import some math functions from the numpy library (note that this</span>
<span class="c1"># isn&#39;t part of the python standard library)</span>
<span class="kn">from</span> <span class="nn">numpy</span> <span class="k">import</span> <span class="n">log10</span><span class="p">,</span> <span class="n">average</span>
<span class="c1"># import argv from the sys module to support basic command line</span>
<span class="c1"># control of this script</span>
<span class="kn">from</span> <span class="nn">sys</span> <span class="k">import</span> <span class="n">argv</span>

<span class="c1">#####</span>
<span class="c1"># Start of function definitions</span>
<span class="c1">#####</span>

<span class="k">def</span> <span class="nf">count_differences</span><span class="p">(</span><span class="n">sequence1</span><span class="p">,</span><span class="n">sequence2</span><span class="p">):</span>
     <span class="sd">&quot;&quot;&quot;Count the number of differences between two sequences of the same length</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># confirm that the two sequences are the same length and throw</span>
     <span class="c1"># an error if they aren&#39;t</span>
     <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence2</span><span class="p">),</span> <span class="s2">&quot;Sequences differ in length&quot;</span>
     <span class="c1"># initiate a counter for the number of differences</span>
     <span class="n">result</span> <span class="o">=</span> <span class="mi">0</span>
     <span class="c1"># iterate over the two sequences and count the number of</span>
     <span class="c1"># positions which are not identical</span>
     <span class="k">for</span> <span class="n">base1</span><span class="p">,</span><span class="n">base2</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sequence1</span><span class="p">,</span><span class="n">sequence2</span><span class="p">):</span>
         <span class="k">if</span> <span class="n">base1</span> <span class="o">!=</span> <span class="n">base2</span><span class="p">:</span>
             <span class="c1"># this is a commonly used shortcut for incrementing a count</span>
             <span class="c1"># and is equivalent to the following statement</span>
             <span class="c1"># result = result + 1</span>
             <span class="n">result</span> <span class="o">+=</span> <span class="mi">1</span>
     <span class="k">return</span> <span class="n">result</span>

<span class="k">def</span> <span class="nf">evolve_seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span>
               <span class="n">substitution_probability</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
               <span class="n">mutation_choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">]):</span>
    <span class="sd">&quot;&quot;&quot;Return two child sequences simulating point mutations</span>

<span class="sd">       An error occurs with probability substitution_probability</span>
<span class="sd">        independently at each position of each child sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Generate two lists for storing the resulting sequences</span>
    <span class="n">r1</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">r2</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">range_length</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">**</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">log10</span><span class="p">(</span><span class="n">substitution_probability</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">range_length</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># a point mutation will occur at this position</span>
            <span class="c1"># what&#39;s wrong with the following statement?</span>
            <span class="n">r1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">choice</span><span class="p">(</span><span class="n">mutation_choices</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># no point mutation at this position</span>
            <span class="n">r1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">range_length</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># a point mutation will occur at this position</span>
            <span class="c1"># what&#39;s wrong with the following statement?</span>
            <span class="n">r2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">choice</span><span class="p">(</span><span class="n">mutation_choices</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># no point mutation at this position</span>
            <span class="n">r2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
    <span class="c1"># convert the lists to strings and return them</span>
    <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">r1</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">root_sequence</span><span class="p">,</span><span class="n">generations</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="c1"># initial some values and perform some basic error checking</span>
    <span class="k">assert</span> <span class="n">generations</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Must simulate one or more generations.&quot;</span>
    <span class="c1"># can you simplify the following test?</span>
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">root_sequence</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">base</span> <span class="o">!=</span> <span class="s1">&#39;A&#39;</span> <span class="ow">or</span> <span class="n">base</span> <span class="o">!=</span> <span class="s1">&#39;C&#39;</span> <span class="ow">or</span> <span class="n">base</span> <span class="o">!=</span> <span class="s1">&#39;G&#39;</span> <span class="ow">or</span> <span class="n">base</span> <span class="o">!=</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span>\
         <span class="s2">&quot;Invalid base identified: </span><span class="si">%s</span><span class="s2">. Only A, C, G, or T are allowed.&quot;</span> <span class="o">%</span> <span class="n">base</span>
    <span class="c1"># initialize a list of the previous generations sequences - this gets used</span>
    <span class="c1"># in the for loop below. since we&#39;ll start with the first generation of</span>
    <span class="c1"># children, root_sequence is the previous generation&#39;s sequence</span>
    <span class="n">previous_generation_sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">root_sequence</span><span class="p">]</span>

    <span class="c1"># iterate over each generation (why do we add one to generations?)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">generations</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="c1"># print the generation number and the current number of sequences</span>
        <span class="nb">print</span> <span class="s2">&quot;Generation: </span><span class="si">%d</span><span class="s2"> (Number of child sequences: </span><span class="si">%d</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="mi">2</span><span class="o">**</span><span class="n">i</span><span class="p">)</span>
        <span class="c1"># create a list to store the current generation of sequences</span>
        <span class="n">current_generation_sequences</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># create a list to store the differences in each current generation</span>
        <span class="c1"># sequence from the root sequence</span>
        <span class="n">difference_counts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># iterate over the sequences of the previous generation</span>
        <span class="k">for</span> <span class="n">parent_sequence</span> <span class="ow">in</span> <span class="n">previous_generation_sequences</span><span class="p">:</span>
            <span class="c1"># evolve two child sequences</span>
            <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span> <span class="o">=</span> <span class="n">evolve_seq</span><span class="p">(</span><span class="n">parent_sequence</span><span class="p">)</span>
            <span class="c1"># count the differences in the first sequence (from root_sequence)</span>
            <span class="n">r1_diffs</span> <span class="o">=</span> <span class="n">count_differences</span><span class="p">(</span><span class="n">root_sequence</span><span class="p">,</span><span class="n">r1</span><span class="p">)</span>
            <span class="c1"># append the count of differences to the list of difference counts</span>
            <span class="n">difference_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r1_diffs</span><span class="p">)</span>
            <span class="c1"># add the new sequence to the list of this generation&#39;s sequences</span>
            <span class="n">current_generation_sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r1</span><span class="p">)</span>
            <span class="c1"># count the differences in the second sequence (from root_sequence)</span>
            <span class="n">r2_diffs</span> <span class="o">=</span> <span class="n">count_differences</span><span class="p">(</span><span class="n">root_sequence</span><span class="p">,</span><span class="n">r2</span><span class="p">)</span>
            <span class="c1"># append the count of differences to the list of difference counts</span>
            <span class="n">difference_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r2_diffs</span><span class="p">)</span>
            <span class="c1"># add the new sequence to the list of this generation&#39;s sequences</span>
            <span class="n">current_generation_sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                <span class="c1"># if the caller specified verbose output, print the actual sequences</span>
                <span class="nb">print</span> <span class="s2">&quot;  </span><span class="si">%s</span><span class="s2"> </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">r1</span><span class="p">,</span> <span class="n">r1_diffs</span><span class="p">)</span>
                <span class="nb">print</span> <span class="s2">&quot;  </span><span class="si">%s</span><span class="s2"> </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">r2</span><span class="p">,</span> <span class="n">r2_diffs</span><span class="p">)</span>
        <span class="c1"># print summary information: the average number of differences in the current</span>
        <span class="c1"># generation from root_sequence</span>
        <span class="nb">print</span> <span class="s2">&quot;Mean differences </span><span class="si">%1.3f</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">average</span><span class="p">(</span><span class="n">difference_counts</span><span class="p">)</span>
        <span class="c1"># current_generation_sequences becomes the next generation&#39;s</span>
        <span class="c1"># previous_generation_sequences</span>
        <span class="n">previous_generation_sequences</span> <span class="o">=</span> <span class="n">current_generation_sequences</span>

    <span class="c1"># upon completion of all generations, return the last generation&#39;s sequences</span>
    <span class="k">return</span> <span class="n">previous_generation_sequences</span>

<span class="c1">#####</span>
<span class="c1"># End of function definitions</span>
<span class="c1">#####</span>

<span class="c1">#####</span>
<span class="c1"># Start main execution block</span>
<span class="c1">#####</span>

<span class="n">script_name</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">,</span> <span class="n">num_generations</span><span class="p">,</span> <span class="n">output_filepath</span> <span class="o">=</span> <span class="n">argv</span>

<span class="c1"># generate a random sequence composed of [&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;]</span>
<span class="c1"># of length sequence_length</span>
<span class="n">root_sequence</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">)):</span>
    <span class="n">root_sequence</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">choice</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)))</span>
<span class="n">root_sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">root_sequence</span><span class="p">)</span>

<span class="c1"># run the simulation and get the final generation of sequences</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="n">main</span><span class="p">(</span><span class="n">root_sequence</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="n">num_generations</span><span class="p">))</span>

<span class="c1"># write the final generation of sequences to a fasta file</span>
<span class="n">output_f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filepath</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">s</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequences</span><span class="p">):</span>
    <span class="n">output_f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&gt;seq</span><span class="si">%d</span><span class="se">\n</span><span class="si">%s</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">s</span><span class="p">))</span>
<span class="n">output_f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c1">#####</span>
<span class="c1"># End main execution block</span>
<span class="c1">#####</span>
</pre></div>
</div>
</div>
</div>


      </div>
      <div class="bottomnav" role="navigation" aria-label="bottom navigation">
      
        <p>
        «&#160;&#160;<a href="lecture7.html">Lecture 7: Molecular evolution and introduction to sequence searching and phylogenetic inference</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="lecture9.html">Lecture 9: Sequence alignment algorithms</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, J Gregory Caporaso.
      Last updated on Apr 11, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.1.
    </div>
  </body>
</html>
